File: src/cli/commands/list.ts
/**
 * List command implementation
 * 
 * Lists all host entries from ~/.hosts
 */

import { readHostsFile } from "../../core/file-io";
import type { Entry, Group, HostsFile } from "../../domain/types";

export type ListOptions = {
  json?: boolean;
  hostsFile?: string;
};

type EntryWithGroup = Entry & {
  group: string;
};

/**
 * Execute the list command
 * 
 * @param options - Command options
 * @returns Exit code (0 = success, 2 = I/O error)
 */
export async function listCommand(options: ListOptions = {}): Promise<number> {
  try {
    // Read ~/.hosts
    const hostsFilePath = options.hostsFile || "~/.hosts";
    const hostsFile = await readHostsFile(hostsFilePath);

    // Collect all entries with their group paths
    const entriesWithGroups = collectEntriesWithGroups(hostsFile);

    // Output based on format
    if (options.json) {
      // JSON format
      console.log(JSON.stringify(entriesWithGroups, null, 2));
    } else {
      // Human-readable table format
      printTable(entriesWithGroups);
    }

    return 0;

  } catch (err: any) {
    // Handle I/O errors
    console.error(`Error: ${err.message}`);
    return 2;
  }
}

/**
 * Collect all entries from all groups with their group paths
 */
function collectEntriesWithGroups(hostsFile: HostsFile): EntryWithGroup[] {
  const entries: EntryWithGroup[] = [];
  
  function collectFromGroup(group: Group, parentPath: string) {
    // Build current group path
    const currentPath = parentPath 
      ? (group.name ? `${parentPath}/${group.name}` : parentPath)
      : group.name;
    
    // Add entries from this group
    for (const entry of group.entries) {
      entries.push({
        ...entry,
        group: currentPath,
      });
    }
    
    // Recursively collect from nested groups
    for (const subgroup of group.groups) {
      collectFromGroup(subgroup, currentPath);
    }
  }
  
  for (const group of hostsFile.groups) {
    collectFromGroup(group, "");
  }
  
  return entries;
}

/**
 * Print entries in human-readable table format
 */
function printTable(entries: EntryWithGroup[]) {
  if (entries.length === 0) {
    console.log("No entries found.");
    return;
  }

  // Print header
  console.log("Status  IP              Hostname                 Aliases          Group");
  console.log("------  --------------  -----------------------  ---------------  -----");

  // Print each entry
  for (const entry of entries) {
    const status = entry.enabled ? "✓" : "✗";
    const ip = entry.ip.padEnd(14);
    const hostname = entry.hostname.padEnd(23);
    const aliases = entry.aliases.join(", ").padEnd(15);
    const group = entry.group || "(ungrouped)";
    
    console.log(`${status}       ${ip}  ${hostname}  ${aliases}  ${group}`);
  }
}

File: src/internal/cmd/list.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/hungthai1401/hostie/src/internal/core/fileio"
	"github.com/hungthai1401/hostie/src/internal/domain"
	"github.com/spf13/cobra"
)

var (
	listJSON bool
)

func newListCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List all hosts entries",
		Long:  "List all entries in the hosts file with optional JSON output",
		Args:  cobra.NoArgs,
		RunE:  runList,
	}

	cmd.Flags().BoolVar(&listJSON, "json", false, "output in JSON format")

	return cmd
}

func runList(cmd *cobra.Command, args []string) error {
	// Read existing hosts file
	hostsFile, err := fileio.ReadHostsFile(hostsFilePath)
	if err != nil {
		return fmt.Errorf("failed to read hosts file: %w", err)
	}

	if listJSON {
		return outputJSON(hostsFile)
	}

	return outputHuman(hostsFile)
}

func outputJSON(hf domain.HostsFile) error {
	// Flatten all entries from all groups (including nested groups)
	entries := flattenEntries(hf.Groups)

	// Create output format matching v1: array of entry objects with lowercase keys
	type JSONEntry struct {
		ID       string   `json:"id"`
		IP       string   `json:"ip"`
		Hostname string   `json:"hostname"`
		Aliases  []string `json:"aliases"`
		Enabled  bool     `json:"enabled"`
		Comment  string   `json:"comment,omitempty"`
		Group    string   `json:"group"`
	}

	var output []JSONEntry
	for _, e := range entries {
		output = append(output, JSONEntry{
			ID:       e.Entry.ID,
			IP:       e.Entry.IP,
			Hostname: e.Entry.Hostname,
			Aliases:  e.Entry.Aliases,
			Enabled:  e.Entry.Enabled,
			Comment:  e.Entry.Comment,
			Group:    e.GroupPath,
		})
	}

	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(output); err != nil {
		return fmt.Errorf("failed to encode JSON: %w", err)
	}
	return nil
}

// flattenEntries recursively flattens all entries from groups and subgroups.
type flatEntry struct {
	Entry     domain.Entry
	GroupPath string
}

func flattenEntries(groups []domain.Group) []flatEntry {
	var result []flatEntry
	for _, g := range groups {
		flattenGroup(&g, g.Name, &result)
	}
	return result
}

func flattenGroup(g *domain.Group, path string, result *[]flatEntry) {
	for _, e := range g.Entries {
		*result = append(*result, flatEntry{
			Entry:     e,
			GroupPath: path,
		})
	}
	for i := range g.Groups {
		subPath := path + "/" + g.Groups[i].Name
		flattenGroup(&g.Groups[i], subPath, result)
	}
}

func outputHuman(hf domain.HostsFile) error {
	if len(hf.Groups) == 0 {
		fmt.Println("No entries found")
		return nil
	}

	for _, group := range hf.Groups {
		printGroup(&group, 0)
	}

	return nil
}

func printGroup(g *domain.Group, indent int) {
	prefix := ""
	for i := 0; i < indent; i++ {
		prefix += "  "
	}

	// Print group header
	if g.Name != "" {
		fmt.Printf("%s[%s]\n", prefix, g.Name)
		if g.Description != "" {
			fmt.Printf("%s  %s\n", prefix, g.Description)
		}
	}

	// Print entries
	for _, entry := range g.Entries {
		status := "✓"
		if !entry.Enabled {
			status = "✗"
		}

		aliasStr := ""
		if len(entry.Aliases) > 0 {
			aliasStr = fmt.Sprintf(" (aliases: %v)", entry.Aliases)
		}

		commentStr := ""
		if entry.Comment != "" {
			commentStr = fmt.Sprintf(" # %s", entry.Comment)
		}

		fmt.Printf("%s  %s %s → %s%s%s\n", prefix, status, entry.Hostname, entry.IP, aliasStr, commentStr)
	}

	// Print subgroups
	for i := range g.Groups {
		printGroup(&g.Groups[i], indent+1)
	}
}
